solana/bridge/program/tests/integration.rs

```diff
@@ -325,53 +382,91 @@ fn test_invalid_emitter(context: &mut Context) {
             instruction,
         ],
-        solana_sdk::commitment_config::CommitmentConfig::processed(),
+        solana_sdk::commitment_config::CommitmentLevel::Processed,
     )
+    .await
     .is_err());
 }
 
-fn test_duplicate_messages_fail(context: &mut Context) {
-    let (ref payer, ref client, ref program) = common::setup();
+#[tokio::test]
+async fn duplicate_messages_fail() {
+    let (ref mut context, ref mut client, ref payer, ref program) = initialize().await;
 
     // We'll use the following nonce/message/emitter/sequence twice.
     let nonce = rand::thread_rng().gen();
-    let message = [0u8; 32].to_vec();
+    let data = [0u8; 32].to_vec();
     let emitter = Keypair::new();
     let sequence = context.seq.next(emitter.pubkey().to_bytes());
+    let fee = 10_000;
+
+    // `common::post_message` creates a new message account every time so we need to manually invoke
+    // the instruction here. Otherwise the account derivation for the second call will be different
+    // and will not fail.
+
+    let fee_collector = FeeCollector::<'_>::key(None, program);
+    let message = Keypair::new();
+    let instruction = instructions::post_message(
+        *program,
+        payer.pubkey(),
+        emitter.pubkey(),
+        message.pubkey(),
+        nonce,
+        data.clone(),
+        ConsistencyLevel::Confirmed,
+    )
+    .unwrap();
 
     // Post the message, publishing the data for guardian consumption.
-    let message_key = common::post_message(
+    common::execute(
         client,
-        program,
         payer,
-        &emitter,
-        nonce,
-        message.clone(),
-        10_000,
+        &[payer, &emitter, &message],
+        &[
+            system_instruction::transfer(&payer.pubkey(), &fee_collector, fee),
+            instruction.clone(),
+        ],
+        CommitmentLevel::Processed,
     )
+    .await
     .unwrap();
 
+    let posted_message: PostedVAAData = common::get_account_data(client, message.pubkey()).await;
+    assert_eq!(posted_message.0.vaa_version, 0);
+    assert_eq!(posted_message.0.consistency_level, 1);
+    assert_eq!(posted_message.0.nonce, nonce);
+    assert_eq!(posted_message.0.sequence, sequence);
+    assert_eq!(posted_message.0.emitter_chain, 1);
+    assert_eq!(&posted_message.0.emitter_address, emitter.pubkey().as_ref());
+    assert_eq!(posted_message.0.payload, data);
+    assert_eq!(
+        posted_message.0.emitter_address,
+        emitter.pubkey().to_bytes()
+    );
+
     // Second should fail due to duplicate derivations.
-    assert!(common::post_message(
+    assert!(common::execute(
         client,
-        program,
         payer,
-        &emitter,
-        nonce,
-        message.clone(),
-        10_000,
+        &[payer, &emitter, &message],
+        &[
+            system_instruction::transfer(&payer.pubkey(), &fee_collector, fee),
+            instruction,
+        ],
+        CommitmentLevel::Processed,
     )
+    .await
     .is_err());
```

Comment:
I had to re-write this test and it's very likely I've screwed something up.  Like I mentioned in the commit message, this test fails sometimes when all tests are run in parallel but hasn't failed a single time when run by itself or with `--test-threads 1`.

The test failure is always the same.  The second execution of the instructions succeeds when it shouldn't.  The weird thing is that it should fail because the message account is already initialized, which should cause the derivation to fail and lines 438 - 449 check the contents of the message account to ensure that it does exist but somehow the second execution still succeeds.

If anyone has thoughts on what might be going on here, I'd love to hear them.

Reply:
The message account is indeed already initialised... what's going on? 🤔

Reply:
As discussed offline this test was wrong so I've removed it.  I didn't spend more time trying to figure out why it was only failing in multi-threaded mode.